### PR TITLE
test: [cpp] add cb explore simulator for unit testing

### PR DESCRIPTION
**Pull request by bassmang**

### CODE REVIEW

File: test/unit_test/tutorial_test.cc [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+#include <fmt/format.h>
+
+class Simulator
+{
+  float USER_LIKED_ARTICLE = -1.f;
+  float USER_DISLIKED_ARTICLE = 0.f;
+  std::vector<std::string> users{"Tom", "Anna"};
+  std::vector<std::string> times_of_day{"morning", "afternoon"};
+  std::vector<std::string> actions{"politics", "sports", "music", "food", "finance", "health", "camping"};
+  std::string debug_logfile;
```
Comment (lalo) [inline test/unit_test/tutorial_test.cc:20]: remove debug_logfile, no need for it right now.

Comment (bassmang) [reply test/unit_test/tutorial_test.cc]: Yea wasnt sure if we might want it in the future, just took it out

File: test/unit_test/tutorial_test.cc [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+      if (context["time_of_day"] == "morning" && action == "sports") { return USER_LIKED_ARTICLE; }
+      else if (context["time_of_day"] == "afternoon" && action == "politics")
+      {
+        return USER_LIKED_ARTICLE;
+      }
+    }
+    return USER_DISLIKED_ARTICLE;
+  }
+
+  std::vector<std::string> to_vw_example_format(
```
Comment (lalo) [inline test/unit_test/tutorial_test.cc:48]:
```suggestion
  // todo: skip text format and create vw example directly
  std::vector<std::string> to_vw_example_format(
```

File: test/unit_test/tutorial_test.cc [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+#include <boost/test/unit_test.hpp>
+#include <boost/test/test_tools.hpp>
+
+#include "vw.h"
+#include "test_common.h"
+#include <numeric>
+#include <fmt/format.h>
+
+class Simulator
```
Comment (lalo) [inline test/unit_test/tutorial_test.cc:13]: should this class live on its own file and namespace? this way we can re-use in other unit tests.

**Review by lalo — APPROVED** [review]

File: test/unit_test/simulator.cc [hunk 4]
```diff
@@ -0,0 +1,127 @@
+#include "simulator.h"
+
+namespace simulator
+{
+cb_sim::cb_sim(int seed) : seed(seed) { srand(seed); }
+
+float cb_sim::get_cost(std::map<std::string, std::string> context, std::string action)
```
Comment (jackgerrits) [inline test/unit_test/simulator.cc:7]:
```suggestion
float cb_sim::get_cost(const std::map<std::string, std::string>& context, std::string action)
```

File: test/unit_test/simulator.cc [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    }
+  }
+  return USER_DISLIKED_ARTICLE;
+}
+// todo: skip text format and create vw example directly
+std::vector<std::string> cb_sim::to_vw_example_format(
+    std::map<std::string, std::string> context, std::string chosen_action, float cost, float prob)
+{
+  std::vector<std::string> multi_ex_str;
+  multi_ex_str.push_back(fmt::format("shared |User user={} time_of_day={}", context["user"], context["time_of_day"]));
```
Comment (jackgerrits) [inline test/unit_test/simulator.cc:32]: Ah its nice to be able to use fmt :)

File: test/unit_test/simulator.cc [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    else if (context["time_of_day"] == "afternoon" && action == "politics")
+    {
+      return USER_LIKED_ARTICLE;
+    }
+  }
+  return USER_DISLIKED_ARTICLE;
+}
+// todo: skip text format and create vw example directly
+std::vector<std::string> cb_sim::to_vw_example_format(
+    std::map<std::string, std::string> context, std::string chosen_action, float cost, float prob)
```
Comment (jackgerrits) [inline test/unit_test/simulator.cc:29]:
```suggestion
    const std::map<std::string, std::string>& context, const std::string& chosen_action, float cost, float prob)
```

File: test/unit_test/simulator.cc [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+  {
+    std::ostringstream ex;
+    if (action == chosen_action) { ex << fmt::format("0:{}:{} ", cost, prob); }
+    ex << fmt::format("|Action article={}", action);
+    multi_ex_str.push_back(ex.str());
+  }
+  return multi_ex_str;
+}
+
+std::pair<int, float> cb_sim::sample_custom_pmf(std::vector<float> pmf)
```
Comment (jackgerrits) [inline test/unit_test/simulator.cc:43]:
```suggestion
std::pair<int, float> cb_sim::sample_custom_pmf(const std::vector<float>& pmf)
```

File: test/unit_test/simulator.cc [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+{
+  float total = std::accumulate(pmf.begin(), pmf.end(), 0.f);
+  float scale = 1.f / total;
+  for (float& val : pmf) { val *= scale; }
+  float draw = static_cast<float>(rand()) / static_cast<float>(RAND_MAX);
+  float sum_prob = 0.f;
+  for (int index = 0; index < pmf.size(); ++index)
+  {
+    sum_prob += pmf[index];
+    if (sum_prob > draw) { return std::pair<int, float>(index, pmf[index]); }
```
Comment (jackgerrits) [inline test/unit_test/simulator.cc:53]:
```suggestion
    if (sum_prob > draw) { return std::make_pair(index, pmf[index]); }
```

File: test/unit_test/simulator.cc [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+  }
+  THROW("Error: No prob selected");
+}
+
+std::pair<std::string, float> cb_sim::get_action(vw* vw, std::map<std::string, std::string> context)
```
Comment (jackgerrits) [inline test/unit_test/simulator.cc:58]:
```suggestion
std::pair<std::string, float> cb_sim::get_action(vw* vw, const std::map<std::string, std::string>& context)
```

File: test/unit_test/simulator.cc [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+
+  std::vector<float> pmf;
+  auto const& scores = examples[0]->pred.a_s;
+  std::vector<float> ordered_scores(scores.size());
+  for (auto const& action_score : scores) { ordered_scores[action_score.action] = action_score.score; }
+  for (auto action_score : ordered_scores) { pmf.push_back(action_score); }
+  vw->finish_example(examples);
+
+  std::pair<int, float> pmf_sample = sample_custom_pmf(pmf);
+  return std::pair<std::string, float>(actions[pmf_sample.first], pmf_sample.second);
```
Comment (jackgerrits) [inline test/unit_test/simulator.cc:73]:
```suggestion
  return std::make_pair(actions[pmf_sample.first], pmf_sample.second);
```

File: test/unit_test/simulator.cc [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+std::string cb_sim::choose_user()
```
Comment (jackgerrits) [inline test/unit_test/simulator.cc:76]:
```suggestion
const std::string& cb_sim::choose_user()
```

File: test/unit_test/simulator.cc [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+{
+  int rand_ind = rand() % users.size();
+  return users[rand_ind];
+}
+
+std::string cb_sim::choose_time_of_day()
```
Comment (jackgerrits) [inline test/unit_test/simulator.cc:82]:
```suggestion
const std::string& cb_sim::choose_time_of_day()
```

File: test/unit_test/tutorial_test.cc [hunk 13]
```diff
@@ -0,0 +1,57 @@
+#ifndef STATIC_LINK_VW
+#  define BOOST_TEST_DYN_LINK
+#endif
+
+#include <boost/test/unit_test.hpp>
+#include <boost/test/test_tools.hpp>
+
+#include "simulator.h"
+
+std::vector<float> _test_helper(std::string vw_arg, int num_iterations = 3000, int seed = 10)
```
Comment (jackgerrits) [inline test/unit_test/tutorial_test.cc:10]:
```suggestion
std::vector<float> _test_helper(const std::string& vw_arg, int num_iterations = 3000, int seed = 10)
```

File: test/unit_test/tutorial_test.cc [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+{
+  auto vw = VW::initialize(vw_arg);
+  simulator::cb_sim sim(seed);
+  auto ctr = sim.run_simulation(vw, num_iterations);
+  VW::finish(*vw);
+  return ctr;
+}
+
+std::vector<float> _test_helper_save_load(std::string vw_arg, int num_iterations = 3000, int seed = 10)
```
Comment (jackgerrits) [inline test/unit_test/tutorial_test.cc:19]:
```suggestion
std::vector<float> _test_helper_save_load(const std::string& vw_arg, int num_iterations = 3000, int seed = 10)
```

File: test/unit_test/simulator.cc (hunk 7 above)
Comment (bassmang) [reply test/unit_test/simulator.cc]: This actually changes the values of pmf so it can't be `const`. Let me know if all the other changes look good

**Review by jackgerrits — APPROVED** [review]

**Comment by lalo** [comment]
related https://github.com/VowpalWabbit/vowpal_wabbit/pull/3252

**Pull request merged**